### PR TITLE
Multicore Non-Negative Least Squares Code and Tests

### PR DESCRIPTION
**Pull request by tofstie**
# Multicore Non-Negative Least Squares Code and Tests
Pull request for the addition of allowing the Non-Negative Least Squares (NNLS) to be run on multiple cores.
## NNLS_solver.cpp

- The input matrix and vector can now exist within multiple cores
- The output matrix will now have the map and distribution of A.DomainMap()
- All calculations will still happen on one core, but this adds the functionality of running NNLS with more than 1 core, speeding up other areas of the code.

## helper_functions.cpp

- Fixed the epetra_to_eig_vec function to properly output the Eigen::VectorXd to all cores
- Added eig_to_epetra_vec function to output an Epetra_Vector from data in a Eigen::VectorXd

## 2D_inviscid_isentropic_vortex.prm
- Corrected the interpolate initial condition parameter

## NNLS_tests.cpp
- Implemented new helper_function
- Added new test for multicore NNLS

### CODE REVIEW

**Comment by tofstie** [comment]
Here are the failed tests from running ctest: [LastTestsFailed.log](https://github.com/user-attachments/files/16416906/LastTestsFailed.log)
All of them exist in the Issues tab under the label testfail

File: src/linear_solver/NNLS_solver.cpp
Comment (cpethrick) [inline src/linear_solver/NNLS_solver.cpp:206]: It's good practice to add `const` when possible.

File: src/linear_solver/NNLS_solver.h
Comment (cpethrick) [inline src/linear_solver/NNLS_solver.h:137]: Could you make sure that the documentation comments in this class comply with the Doxygen conventions of `/// comment` on the line before or `///<` immediately following a variable?

File: tests/unit_tests/linear_solver/CMakeLists.txt
Comment (cpethrick) [inline tests/unit_tests/linear_solver/CMakeLists.txt:59]: Is there a reason why you use `/usr/bin/mpiexec -np 2` rather than `mpirun -n ${MPIMAX} ` ?

File: tests/unit_tests/linear_solver/NNLS_tests.cpp
Comment (cpethrick) [inline tests/unit_tests/linear_solver/NNLS_tests.cpp:670]: I'm not familiar with this part of the code, but can you explain why you're able to remove `b_pt`?

**Review by cpethrick — COMMENTED** [review]
Thanks for the PR, multicore functionality will be very useful for the ROM part of PHiLiP! 

I've left a few comments which should be quick to resolve. Could you also confirm whether the result is the same when you run a test on a single processor vs. multiple processors?

File: tests/unit_tests/linear_solver/NNLS_tests.cpp
Comment (tofstie) [reply tests/unit_tests/linear_solver/NNLS_tests.cpp:670]: Of course, b_pt was only used to construct Epetra_Vector b using the following line
`Epetra_Vector b(Copy, A.RowMap(), b_pt);`
However, if A.RowMap() streches over more than one core, it will only copy the number of local elements from b_pt starting from 0. The resulting vector will have the following data and format. 
<code>Epetra::Vector\
Vector b\
MyPID&nbsp;GID&nbsp;Value\
0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.6\
0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.2\
1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.6\
1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.2
</code>
To correct this, the helper function eig_to_epetra_vector corrects this issue to give the following Epetra_Vector. b_pt is no longer needed and was deleted from NNLS_tests.cpp.
<code>Epetra::Vector\
Vector b\
MyPID&nbsp;GID&nbsp;Value\
0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.6\
0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.2\
1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.8\
1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;8.4
</code>

File: src/linear_solver/NNLS_solver.h
Comment (tofstie) [reply src/linear_solver/NNLS_solver.h:139]: Completed in latest commit for all variables in NNLS_solver.h

File: tests/unit_tests/linear_solver/CMakeLists.txt
Comment (tofstie) [reply tests/unit_tests/linear_solver/CMakeLists.txt]: There was not, I just forgot to change it in the last commit. 
I have changed it to `mpirun -n ${MPIMAX}` in the latest commit

**Comment by tofstie** [comment]
> Thanks for the PR, multicore functionality will be very useful for the ROM part of PHiLiP!
> 
> I've left a few comments which should be quick to resolve. Could you also confirm whether the result is the same when you run a test on a single processor vs. multiple processors?

The results are the same for a single processor and multiple processors. I will use Case 1 in the unit tests as an example. 
### Single Core
<code>Epetra::Vector\
Solution x\
MyPID&nbsp;GID&nbsp;Value\
0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.1\
0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.5\
</code>
### 2 Cores
<code>Epetra::Vector\
Solution x\
MyPID&nbsp;GID&nbsp;Value\
0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.1\
1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.5\
</code>
The only difference in the two vectors is where the values are stored. This is because there is no change to the logic of the NNLS solver, just that the input and output can be Epetra objects that are distributed over multiple cores.

File: src/linear_solver/NNLS_solver.cpp
Comment (tofstie) [reply src/linear_solver/NNLS_solver.cpp]: Changed this line and a few others in the following functions
`NNLS_solver::allocateVectorSingleCore(const Epetra_Vector &b)`
`NNLS_solver::allocateMatrixSingleCore(const Epetra_CrsMatrix &A)`
`NNLS_solver::allocateVectorSingleCore(const Epetra_Vector &c)`

File: src/linear_solver/NNLS_solver.cpp
Comment (PranshulThakur) [inline src/linear_solver/NNLS_solver.cpp:382]: This function is returning a matrix which will then be copied into another matrix. Consider passing in the output matrix as a reference and then updating it within the function it to improve efficiency.

File: src/linear_solver/NNLS_solver.h
Comment (PranshulThakur) [inline src/linear_solver/NNLS_solver.h:196]: Renaming to `allocateMatrixToSingleCore` and `allocateVectortoSingleCore` would be more descriptive.

File: src/linear_solver/helper_functions.cpp
Comment (PranshulThakur) [inline src/linear_solver/helper_functions.cpp:22]: Might be better to use `//` for comments which are not meant to appear in doxygen.

**Review by PranshulThakur — COMMENTED** [review]
Amazing first PR to run NNLS on multiple cores. The changes involve a few hundred lines, which is perfect and makes it easier to review. Great job! I have a few comments and will approve it soon.

File: tests/unit_tests/linear_solver/CMakeLists.txt
Comment (cpethrick) [reply tests/unit_tests/linear_solver/CMakeLists.txt]: Great. Just wondering, if the user is running a lot of cores and there are more cores than the length of the vector in the unit test, would that cause an issue?

File: src/linear_solver/NNLS_solver.h
Comment (cpethrick) [reply src/linear_solver/NNLS_solver.h:139]: Great, thanks!

File: src/linear_solver/NNLS_solver.cpp
Comment (cpethrick) [reply src/linear_solver/NNLS_solver.cpp]: Perfect!

**Comment by cpethrick** [comment]
Thanks for responding to my comments!

Regarding the question about the solution being the same on different numbers of cores, it's something to look out for as you increase the complexity of problems. We've seen it before where there were differences of ~10^-14 when running on different numbers of cores, but I am not sure whether that is in the most recent version of master or not. I'm glad that you get the same result for your unit test!

Approving, we can merge it in once @PranshulThakur and @biondicc have looked it over.

**Review by cpethrick — APPROVED** [review]

File: src/linear_solver/NNLS_solver.cpp
Comment (PranshulThakur) [reply src/linear_solver/NNLS_solver.cpp:382]: As discussed with @tofstie , since this function is used to initialize matrices in the constructor, it's better for it to return an Epetra matrix. Marking this comment as resolved.

File: src/linear_solver/NNLS_solver.h
Comment (tofstie) [reply src/linear_solver/NNLS_solver.h]: Changed this variable to lowercase. Also changed one other variable that had capital letters.

File: src/linear_solver/NNLS_solver.h
Comment (tofstie) [reply src/linear_solver/NNLS_solver.h]: Done, did for all multi-line Doxygen briefs in commit eb47054.

File: src/linear_solver/NNLS_solver.cpp
Comment (tofstie) [reply src/linear_solver/NNLS_solver.cpp]: I have changed the constructors to call a common constructor with specified values depending on which overloaded constructor is called. This was changed in commit eb47054.

File: src/linear_solver/NNLS_solver.h
Comment (tofstie) [reply src/linear_solver/NNLS_solver.h]: I have renamed all three of the epetra allocation functions to have the more descriptive names in commit eb47054

File: src/linear_solver/helper_functions.cpp
Comment (tofstie) [reply src/linear_solver/helper_functions.cpp]: Changed, I have also scanned the other files changed in this commit to make sure that this issue isn't present in other files.

**Comment by tofstie** [comment]
> Thanks for responding to my comments!
> 
> Regarding the question about the solution being the same on different numbers of cores, it's something to look out for as you increase the complexity of problems. We've seen it before where there were differences of ~10^-14 when running on different numbers of cores, but I am not sure whether that is in the most recent version of master or not. I'm glad that you get the same result for your unit test!
> 
> Approving, we can merge it in once @PranshulThakur and @biondicc have looked it over.

Just to make sure that there were no difference, I have tried the Matlab test case for 1 core and 2 core, which returns a vector of size 1024. I loaded the results into matlab, ensuring that the results are within `std::numeric_limits<double>::digits10`, and the difference of the vectors was exactly the zero vector. If you want to convince yourself, here are the results for 1 Core and 2 Core, with the second number being the rank of the processor.
[2Corex,0.txt](https://github.com/user-attachments/files/16446617/2Corex.0.txt)
[2Corex,1.txt](https://github.com/user-attachments/files/16446618/2Corex.1.txt)
[1Corex,0.txt](https://github.com/user-attachments/files/16446619/1Corex.0.txt)
I also tried between 1 core and 8 core and again the difference had a norm of 0.

**Review by PranshulThakur — COMMENTED** [review]
Thanks @tofstie for addressing all my comments. Can you try running `make doc 2> output_error_doc.log` in the build directory and see if there aren't any doxygen warnings from the functions you have implemented? The PR looks good to me but I will leave it to @biondicc to review and provide the final approval.

File: tests/unit_tests/linear_solver/CMakeLists.txt
Comment (tofstie) [reply tests/unit_tests/linear_solver/CMakeLists.txt]: When a user runs the NNLS with more cores than the length of the vector, some of the Epetra_Vector data stored on a core would just have 0 local elements. It does not cause an issue inside the NNLS solver, as all of the solve is done on the root anyway.

**Comment by tofstie** [comment]
> Thanks @tofstie for addressing all my comments. Can you try running `make doc 2> output_error_doc.log` in the build directory and see if there aren't any doxygen warnings from the functions you have implemented? The PR looks good to me but I will leave it to @biondicc to review and provide the final approval.

There are no doxygen warnings from the functions I have implemented. The only warning for the NNLS area is the following
`/home/tyson/Codes/HYPER_PHiLiP/src/linear_solver/NNLS_solver.h:71: warning: Compound PHiLiP::NNLS_solver is not documented.` However, there is documentaiton, but not right above the NNLS. I'll move some of the documentation to the correct spot to ensure this error goes away. I will wait to commit to see if Calista has any issues first.

**Comment by cpethrick** [comment]
> > Thanks for responding to my comments!
> > Regarding the question about the solution being the same on different numbers of cores, it's something to look out for as you increase the complexity of problems. We've seen it before where there were differences of ~10^-14 when running on different numbers of cores, but I am not sure whether that is in the most recent version of master or not. I'm glad that you get the same result for your unit test!
> > Approving, we can merge it in once @PranshulThakur and @biondicc have looked it over.
> 
> Just to make sure that there were no difference, I have tried the Matlab test case for 1 core and 2 core, which returns a vector of size 1024. I loaded the results into matlab, ensuring that the results are within `std::numeric_limits<double>::digits10`, and the difference of the vectors was exactly the zero vector. If you want to convince yourself, here are the results for 1 Core and 2 Core, with the second number being the rank of the processor. [2Corex,0.txt](https://github.com/user-attachments/files/16446617/2Corex.0.txt) [2Corex,1.txt](https://github.com/user-attachments/files/16446618/2Corex.1.txt) [1Corex,0.txt](https://github.com/user-attachments/files/16446619/1Corex.0.txt) I also tried between 1 core and 8 core and again the difference had a norm of 0. Here is the matlab file if you want to try yourself [comparing_multicore_NNLS.zip](https://github.com/user-attachments/files/16446781/comparing_multicore_NNLS.zip)

Sounds good! Just wanted to make sure this was the case :)

File: tests/integration_tests_control_files/isentropic_vortex_integration/2D_inviscid_isentropic_vortex.prm
Comment (biondicc) [inline tests/integration_tests_control_files/isentropic_vortex_integration/2D_inviscid_isentropic_vortex.prm:1]: Is this change related to the NNLS solver?

**Review by biondicc — CHANGES_REQUESTED** [review]
Left one comment, but other than that looks good to me.

File: tests/integration_tests_control_files/isentropic_vortex_integration/2D_inviscid_isentropic_vortex.prm
Comment (tofstie) [reply tests/integration_tests_control_files/isentropic_vortex_integration/2D_inviscid_isentropic_vortex.prm:1]: No, it was just a parameter file that I noticed had an error on one of the parameters in the flow_solver_param subsection. It was a quick fix so I included it in here.

**Review by biondicc — APPROVED** [review]
Good to go

**Review by biondicc — APPROVED** [review]
LGTM

**Pull request merged**
